engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,812 @@
+		g.normalTables[sourceTable] = postTableStmt
+		return true, optimism.ConflictNone
+	}
+
+	// meet conflict DDL
+	g.normalTables[sourceTable] = prevTableStmt
+	g.conflictTables[sourceTable] = postTableStmt
+
+	// if any conflict happened between conflict DDLs, return error
+	// e.g. tb1: "ALTER TABLE RENAME a TO b", tb2: "ALTER TABLE RENAME c TO d"
+	if !g.noConflictForTables(conflict) {
+		log.L().Error("conflict happened with other conflict tables", zap.Any("source table", sourceTable), zap.String("prevTable", prevTableStmt), zap.String("postTable", postTableStmt))
+		return false, optimism.ConflictDetected
+	}
+
+	if g.noConflictForTables(final) {
+		log.L().Info("all conflict resolved for the DDL", zap.Any("source table", sourceTable), zap.String("prevTable", prevTableStmt), zap.String("postTable", postTableStmt))
+		g.resolveTables()
+		return true, optimism.ConflictNone
+	}
+	log.L().Info("conflict hasn't been resolved", zap.Any("source table", sourceTable), zap.Stringer("prevTable", prevTable), zap.Stringer("postTable", postTable))
+	return false, optimism.ConflictSkipWaitRedirect
+}
+
+// joinTables join tables by tableType.
+func (g *shardGroup) joinTables(tp tableType) (schemacmp.Table, error) {
+	var (
+		joined     schemacmp.Table
+		err        error
+		firstTable = true
+	)
+
+	for sourceTable := range g.normalTables {
+		tableStmt, ok := g.getTableBySourceTable(sourceTable, tp)
+		if !ok || tableStmt == "" {
+			continue
+		}
+		table := genCmpTable(tableStmt)
+
+		if firstTable {
+			joined = table
+			firstTable = false
+			continue
+		}
+		joined, err = joined.Join(table)
+		if err != nil {
+			return joined, err
+		}
+	}
+	return joined, nil
+}
+
+func (g *shardGroup) noConflictForTables(tp tableType) bool {
+	if _, err := g.joinTables(tp); err != nil {
+		return false
+	}
+	if !g.allTableSmaller(tp) {
+		return false
+	}
+	if !g.allTableLarger(tp) {
+		return false
+	}
+	return true
+}
+
+func (g *shardGroup) noConflictWithOneNormalTable(sourceTable metadata.SourceTable, prevTable, postTable schemacmp.Table) bool {
+	for st, ti := range g.normalTables {
+		if st == sourceTable {
+			continue
+		}
+		if ti == "" {
+			continue
+		}
+		t := genCmpTable(ti)
+
+		// judge joined no error
+		joined, err := postTable.Join(t)
+		if err != nil {
+			continue
+		}
+
+		// judge this normal table is smaller(same as allTableSmaller)
+		if _, err = joined.Compare(prevTable); err == nil {
+			continue
+		}
+
+		// judge this normal table is larger(same as allTableLarger)
+		if joined, err = prevTable.Join(t); err != nil {
+			if _, err = t.Compare(postTable); err == nil {
+				return true
+			}
+		}
+		if cmp, err := joined.Compare(postTable); err != nil || cmp < 0 {
+			continue
+		}
+
+		return true
+	}
+	return false
+}
+
+func (g *shardGroup) allTableSmaller(tp tableType) bool {
```

Comment:
it checks all current table structure of conflicting tables are smaller than joined table by tableType? why does it not check `cmp < 0`?

Reply:
just copy from https://github.com/pingcap/tiflow/blob/master/dm/pkg/shardddl/optimism/lock.go#L965-L974
it check all tables doesn't have column x

Reply:
[08ad8c1](https://github.com/pingcap/tiflow/pull/7523/commits/08ad8c1c45c73ef4b60b430273f2da5ae05a6686)